src/Common/Threading.h

```diff
@@ -0,0 +1,270 @@
+#pragma once
+
+#include <base/types.h>
+#include <base/defines.h>
+
+#include <Common/Exception.h>
+
+#ifdef OS_LINUX /// Because of futex
+
+#include <atomic>
+#include <mutex>
+#include <unordered_map>
+#include <memory>
+
+namespace DB
+{
+
+// Scoped object, enabling thread cancellation (cannot be nested)
```

Comment:
As far as I see, "nesting" is possible a la
```
Cancellable c1;
NonCancellable nc;
Cancellable c2;
```

The comment in l. 25 states this better - the same comment should be here too.

(typo: it's "Cancelable" (single l))

Reply:
in constructor of Cancellable I have `reset()` instead of `enable()`. It unconditionally erases a previously held cancellation (if any). Yes, looks like this should be in the comment.

Reply:
Random articles on the Internet say that "cancellable" is the British spelling and "cancelable" is the American spelling. I have no strong opinion on the topic, but it looks like both are possible.